src/main/thrift/parquet.thrift

```diff
@@ -237,6 +237,98 @@ struct SizeStatistics {
    3: optional list<i64> definition_level_histogram;
 }
 
+/**
+ * Interpretation for edges of GEOMETRY logical type, i.e. whether the edge
+ * between points represent a straight cartesian line or the shortest line on
+ * the sphere. It applies to all non-point geometry objects.
+ */
+enum Edges {
+  PLANAR = 0;
+  SPHERICAL = 1;
+}
+
+/**
+ * A custom binary-encoded polygon or multi-polygon to represent a covering of
+ * geometries. For example, it may be a bounding box or an envelope of geometries
+ * when a bounding box cannot be built (e.g. a geometry has spherical edges, or if
+ * an edge of geographic coordinates crosses the antimeridian). In addition, it can
+ * also be used to provide vendor-agnostic coverings like S2 or H3 grids.
+ */
+struct Covering {
+  /**
+   * A type of covering. Currently accepted values: "WKB".
+   */
+  1: required string kind;
+  /**
+   * A payload specific to kind. Below are the supported values:
+   * - WKB: well-known binary of a POLYGON or MULTI-POLYGON that completely
+   *   covers the contents. This will be interpreted according to the same CRS
+   *   and edges defined by the logical type.
+   */
+  2: required binary value;
+}
+
+/**
+ * Bounding box of geometries in the representation of min/max value pair of
+ * coordinates from each axis. Values of Z and M are omitted for 2D geometries.
+ * Filter pushdown on geometries are only safe for planar spatial predicate
+ * but it is recommended that the writer always generates bounding box statistics,
+ * regardless of whether the geometries are planar or spherical.
+ */
+struct BoundingBox {
```

Comment:
For spherical edges, are there any restrictions as to what is considered a valid bounding box?

I would even ask the same for Cartesian edges, although it is less ambiguous here. For example, what happens if xmax is smaller than xmin? Is it up to the reader to decide what to do?

Reply:
@mkaravel 

The bounding box statistics will be always calculated by getting the min/max of X and Y values. For cartesian geometries, this is not a problem at all because there is no antimeridian and orientation issue at all in this case.

For spherical geometries, this is a problem so that's why the spec says that it is not safe to use it for spherical geometries.

In this proposal, (1) there is another type of statistics called covering that attempts to solve this issue. (2) there is another field called `string metadata` which can be used to indicate `orientation (clockwise / counter clockwise)` of the geoparquet metadata: https://github.com/opengeospatial/geoparquet/blob/v1.1.0/format-specs/geoparquet.md?plain=1#L46

However, in this proposal, the main focus is the `geometry` not `geography`. The Geography problem (aka spherical geometries) will be the focus of the next phase.

Reply:
> The bounding box statistics will be always calculated by getting the min/max of X and Y values.

For geometries on a sphere or a spheroid this may no longer produce a bounding box if you restrict to just the vertices of the geometries.

Reply:
> However, in this proposal, the main focus is the geometry not geography. The Geography problem (aka spherical geometries) will be the focus of the next phase.

If a GEOGRAPHY going to be a different logical type? Maybe this not such a bad idea, in which case the current proposal may want to focus just on the Cartesian space and nothing else.

Reply:
While PostGIS uses a distinct type for geography, this is not necessarily a good idea that other formats should reproduce (I think it was due to historical constraints). Any geometry associated to a geographic CRS may be considered "geography". And as said by mkaravel, naive computation of min/max values will not work in the latter case.

I suggest to remove the sentence saying "it is recommended that the writer always generates bounding box statistics, regardless of whether the geometries are planar or spherical". I suggest to even avoid saying that the values are min/max, or said that it is the case in Cartesian coordinate system but not necessarily in other coordinate system.

In particular I suggest to keep open the possibility that "min" > "max". This is the only approach that I know which works with anti-meridian. Bounding boxes in the EPSG database are already expressed that way. Some standards such as GGXF also mandate it. It makes calculation of union and intersection more complicated, but libraries only need to define `union` and `intersection` methods in one place and use it consistently.

I'm not saying that the specification should support "min" > "max" now, just suggesting to keep this door open.

Reply:
> dataset authors should split anything that crosses the antimeridian into two polygons

Not necessarily. This approach introduces other problems (e.g. creates an artificial edge), and "min" > "max" is a good alternative which avoids this problem. Some dataset authors already use the latter (e.g. EPSG) and may not agree that they should change. This approach is also already used by some OGC/ISO standards.

> which would allow Parquet implementations to do filter pushdown without ever having to inspect the crs key

It is not necessary to inspect the CRS key if the specification said that every "min" > "max" cases are wraparound. It is not even necessary to know that the wraparound is 360° or any other value. It works in a truly generic way. The only requirement is to have more advanced implementations of the `contains` and `intersects` methods than the naive ones. This is only mathematics, could be documented in the specification, and would be truly generic (not specific to geographic data).

Reply:
This is not a geometry-specific concept. The idea is to reinterpret the "min" and "max" values so that they are no longer min/max, but "start value" and "end value". Then add the following rules:

* The interior is everything from the start value to the end value, _going in the direction of increasing values_.
* If _start_ > _end_, then we go from _start_ value to positive infinity, wrap around to negative infinity and continue until the _end_ value.

That's all. No geometric concept, no need to know that the wraparound happens at ±180°. It works if all coordinates are inside a consistent range of validity. It may be [−180 … +180]° of longitude, or [0 … 360]° or anything else at user's choice (actually defined by the CRS). As long as all coordinates are guaranteed inside the same range, above algorithm does not need to know that range.

This proposal works for all kinds of wraparound: not only longitudes, but also climatological data (e.g. average temperatures of January, February … December, then back to January with no particular year associated to those months), radar, _etc._ By contrast, if we choose to add a `Geography` type for handling longitude wraparound, then we would also need a `Radar` type for data on a polar coordinate system, a `ClimateCalendar` type for temporal coordinate system, _etc._ This is not scalable.

Above needs are not hypothetical. By coincidence, I'm facing this week the problem of using climatological data in GIS system that do not understand temporal wraparound. Those data are common at WMO, NATO and other organizations. The "min" > "max" proposal addresses the general case. The `Geography` type does not.

Reply:
> The bounding box statistics will be always calculated by getting the min/max of X and Y values. For cartesian geometries, this is not a problem at all because there is no antimeridian and orientation issue at all in this case.

> For spherical geometries, this is a problem so that's why the spec says that it is not safe to use it for spherical geometries.

FYI @dmitrykoval .  We were discussing earlier that some engine still have a way to use bounding box even for spherical edges (there's ways to convert?)

Reply:
I am mostly aligned with @desruisseaux on this one. The one catch here is that if you have a longitude start value of 60° and a longitude end value of -60° then "going to +infinity, then wrapping to -infinity" approach produces a longitude range that is more than 180° and I believe this is no longer a geodesic arc on the sphere/spheroid, no matter what the latitude values are (poles excluded).

Reply:
@desruisseaux can you suggest the exact wording for this part such that it will leave the door open?

Reply:
@szehon-ho In our last discussion I suggested that having numeric min/max values for spherical model is still better than having no stats at all. The limitations discussed by @desruisseaux  and @mkaravel  make sense. From the query engine's perspective, we can still perform pruning, but not only when specific conditions are met, for example when: stored objects do not change meridian signs, there are no poles intersections, etc. In these cases, we can take arc curvature into account when evaluating spatial predicates and skip processing of files/rowgroups where no ambiguity exists.

To summarize, min/max stats still make sense compared to no stats, but there are certainly better alternatives.

Reply:
BigQuery (that only supports spherical geographies) supports `ST_IntersectsBox(geo, lng1, lat1, lng2, lat2)` function, and uses it for partition pruning. The conventions match ones described above: lng1 is assumed to be **westmost point**, and lng2 to be **eastmost point**, rather than lng1 being less than lng2. Thus it allows describing boxes that cross anti-meridian, e.g. `ST_IntersectsBox(ST_GeogPoint(180, 0), 170, -10, -170, 10)` can also be written as `ST_IntersectsBox(ST_GeogPoint(180, 0), 170, -10, 190, 10)` and returns TRUE.

I also blogged about our semantics: https://medium.com/geospatial-analytics/squaring-the-sphere-974d1e4a875e

Reply:
As for covering, we discussed using S2 cell list for covering, and there is a proposal how to do partitioning transform using S2: https://docs.google.com/document/d/1tG13UpdNH3i0bVkjFLsE2kXEXCuw1XRpAC2L2qCUox0/edit?usp=sharing but this is probably longer-term solution.

Reply:
Another consideration here is that Covering is more computationally expensive to compute, compared to a bounding box. I can imagine that in some write-performance-sensitive  scenarios, simpler bbox stats would be the preferred approach. Are there any concerns with adopting @mentin's west-east convention for spherical edges? 

@mentin thanks for sharing the S2 transform proposal - will have a look.